libbpf-tools/klockstat.bpf.c

```diff
@@ -0,0 +1,265 @@
+	if (li->stack_id < 0)
+		return;
+	li->try_at = bpf_ktime_get_ns();
+
+	di->task_id = task_id;
+	di->depth = ti->depth;
+	bpf_map_update_elem(&lockholder_map, di, li, BPF_ANY);
+}
+
+static void lock_acquired(struct mutex *lock)
+{
+	u64 task_id;
+	struct task_info *ti;
+	struct lockholder_info *li;
+	struct depth_id di[1] = {0};
+
+	if (targ_lock && targ_lock != lock)
+		return;
+	task_id = bpf_get_current_pid_tgid();
+	if (!tracing_task(task_id))
+		return;
+	ti = bpf_map_lookup_elem(&task_map, &task_id);
+	if (!ti)
+		return;
+	if (!ti->depth)
+		return;
+	di->task_id = task_id;
+	di->depth = ti->depth;
+	li = bpf_map_lookup_elem(&lockholder_map, di);
+	if (!li)
+		return;
+
+	li->acq_at = bpf_ktime_get_ns();
+}
+
+static void account(struct lockholder_info *li)
+{
+	struct lock_stat *ls;
+	u64 delta;
+
+	ls = bpf_map_lookup_elem(&stat_map, &li->stack_id);
+	if (!ls) {
+		struct lock_stat fresh = {0};
+
+		bpf_map_update_elem(&stat_map, &li->stack_id, &fresh, BPF_ANY);
+		ls = bpf_map_lookup_elem(&stat_map, &li->stack_id);
+		if (!ls)
+			return;
+	}
+
+	delta = li->acq_at - li->try_at;
+	ls->acq_count++;
+	ls->acq_total_time += delta;
+	if (delta > ls->acq_max_time) {
+		ls->acq_max_time = delta;
+		ls->acq_max_id = li->task_id;
+		bpf_get_current_comm(ls->acq_max_comm, TASK_COMM_LEN);
+	}
+
+	delta = li->rel_at - li->acq_at;
+	ls->hld_count++;
+	ls->hld_total_time += delta;
+	if (delta > ls->hld_max_time) {
+		ls->hld_max_time = delta;
+		ls->hld_max_id = li->task_id;
+		bpf_get_current_comm(ls->hld_max_comm, TASK_COMM_LEN);
+	}
+}
+
+static void lock_released(struct mutex *lock)
+{
+	u64 task_id;
+	struct task_info *ti;
+	struct lockholder_info *li;
+	struct depth_id di[1] = {0};
+
+	if (targ_lock && targ_lock != lock)
+		return;
+	task_id = bpf_get_current_pid_tgid();
+	if (!tracing_task(task_id))
+		return;
+	ti = bpf_map_lookup_elem(&task_map, &task_id);
+	if (!ti)
+		return;
+	if (!ti->depth)
+		return;
+	di->task_id = task_id;
+	di->depth = ti->depth;
+	li = bpf_map_lookup_elem(&lockholder_map, di);
+	if (!li)
+		return;
+
+	li->rel_at = bpf_ktime_get_ns();
+	account(li);
+
+	bpf_map_delete_elem(&lockholder_map, di);
+
+	ti->depth--;
+}
+
+static void task_exiting(u64 task_id)
+{
+	bpf_map_delete_elem(&task_map, &task_id);
+}
+
+SEC("fentry/mutex_lock")
+int BPF_PROG(mutex_lock, struct mutex *lock)
+{
+	lock_contended(ctx, lock);
+	return 0;
+}
+
+SEC("fexit/mutex_lock")
+int BPF_PROG(mutex_lock_exit, struct mutex *lock, long ret)
+{
+	lock_acquired(lock);
+	return 0;
+}
+
+SEC("fentry/mutex_unlock")
+int BPF_PROG(mutex_unlock, struct mutex *lock)
+{
+	lock_released(lock);
+	return 0;
+}
+
+SEC("tp_btf/sched_switch")
+int BPF_PROG(sched_switch, bool preempt, struct task_struct *prev,
+	     struct task_struct *next)
+{
+	#define TASK_DEAD 0x0080
+	/*
+	 * Need CORE_READ in a conditional for now:
+	 * https://lore.kernel.org/bpf/20200630171240.2523722-1-yhs@fb.com/
+	 */
+	if (BPF_CORE_READ(prev, state) != TASK_DEAD)
```

Comment:
Please use https://github.com/iovisor/bcc/blob/master/libbpf-tools/core_fixes.bpf.h#L20 here.

Reply:
I create a new PR #3691 to resolve what you concern here. So, no worry.

For the two questions:
* We include `vmlinux.h` which has `preserve_access_index` applied and we use that in `fentry` context, which allows kernel pointer chasing, so it's OK.
* This is compiler(clang) and loader(libbpf) magic. From libbpf's perspective, `struct task_struct` and `struct task_struct___x ` are the same, and `bpf_core_field_exists` is not a function call.

Please refer to Andrii's blog: https://nakryiko.com/posts/bpf-core-reference-guide/